nixos/modules/system/boot/systemd/initrd.nix

```diff
@@ -0,0 +1,342 @@
+    boot.initrd.systemd = {
+      emergencyPackages = [pkgs.bash pkgs.coreutils pkgs.kmod cfg.package];
+
+      objects = [
+        { object = "${cfg.package}/lib/systemd/systemd"; symlink = "/init"; }
+        { object = stage1Units; symlink = "/etc/systemd/system"; }
+
+        # TODO: Limit this to the bare necessities
+        { object = "${cfg.package}/lib"; }
+
+        { object = "${cfg.package.util-linux}/bin/mount"; }
+        { object = "${cfg.package.util-linux}/bin/umount"; }
+        { object = "${cfg.package.util-linux}/bin/sulogin"; }
+
+        # TODO: Not sure why this needs to be here for the recovery shell to work
+        { object = "${pkgs.glibc}/lib/libnss_files.so"; }
+
+        { object = config.environment.etc.os-release.source; symlink = "/etc/initrd-release"; }
+        { object = config.environment.etc.os-release.source; symlink = "/etc/os-release"; }
+        { object = fstab; symlink = "/etc/fstab"; }
+        {
+          object = "${modulesClosure}/lib/modules";
+          symlink = "/lib/modules";
+        }
+        {
+          symlink = "/etc/modules-load.d/nixos.conf";
+          object = pkgs.writeText "nixos.conf"
+            (lib.concatStringsSep "\n" config.boot.initrd.kernelModules);
+        }
+        {
+          object = builtins.toFile "passwd" "root:x:0:0:System Administrator:/root:/bin/bash";
+          symlink = "/etc/passwd";
+        }
+        {
+          object = builtins.toFile "shadow" "root:${config.boot.initrd.systemd.emergencyHashedPassword}:::::::";
+          symlink = "/etc/shadow";
+        }
+        { object = "${emergencyEnv}/bin"; symlink = "/bin"; }
+        { object = "${emergencyEnv}/sbin"; symlink = "/sbin"; }
+        { object = builtins.toFile "bashrc" "PATH=/bin:/sbin"; symlink = "/etc/bashrc"; }
+        { object = builtins.toFile "sysctl.conf" "kernel.modprobe = /sbin/modprobe"; symlink = "/etc/sysctl.d/nixos.conf"; }
+      ];
+
+      targets.initrd.aliases = ["default.target"];
+      units =
+           mapAttrs' (n: v: nameValuePair "${n}.path"    (pathToUnit          n v)) cfg.paths
+        // mapAttrs' (n: v: nameValuePair "${n}.service" (initrdServiceToUnit n v)) cfg.services
+        // mapAttrs' (n: v: nameValuePair "${n}.slice"   (sliceToUnit         n v)) cfg.slices
+        // mapAttrs' (n: v: nameValuePair "${n}.socket"  (socketToUnit        n v)) cfg.sockets
+        // mapAttrs' (n: v: nameValuePair "${n}.target"  (targetToUnit        n v)) cfg.targets
+        // mapAttrs' (n: v: nameValuePair "${n}.timer"   (timerToUnit         n v)) cfg.timers
+        // listToAttrs (map
+                     (v: let n = escapeSystemdPath v.where;
+                         in nameValuePair "${n}.mount" (mountToUnit n v)) cfg.mounts)
+        // listToAttrs (map
+                     (v: let n = escapeSystemdPath v.where;
+                         in nameValuePair "${n}.automount" (automountToUnit n v)) cfg.automounts);
```

Comment:
This feels pretty repetitive but it does keep the code simpler vs. a higher level approach. Not sure what to make of that.

Reply:
Yea it's not the first time this is repeated in nixos. I suppose I could factor it out into a function. Would that be good?

Reply:
Yeah, that'd be nice! The generic-unit-ification being in what is effectively the consumer of the systemd lib seems weird.